docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+     from an accessible image repository.
+
+**Example Usage**
+
+As an operator, I would like to fetch my azure provider components from within
+the cluster because I’m working within an air-gapped environment.
+
+In this example, we have two config maps that define the components and
+metadata of the provider. They each share the label `provider-components:
+azure` and are within the `capz-system` namespace.
+
+The azure InfrastructureProvider has a `fetchConfig` which specifies the label
+selector. This way the operator knows which versions of the azure provider are
+available. Since the provider’s version is marked as `v0.4.9`, it uses the
+components information from the config map to install the azure provider.
+
+```yaml
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ labels:
+   provider-components: azure
+ name: v0.4.9
+ namespace: capz-system
+data:
+ components: |
+ # components for v0.4.9 yaml goes here
+ metadata: |
+ # metadata information goes here
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ labels:
+   provider-components: azure
+ name: v0.4.8
+ namespace: capz-system
+data:
+ components: |
+ # components for v0.4.8 yaml goes here
+ metadata: |
+ # metadata information goes here
+---
+apiVersion: management.cluster.x-k8s.io/v1alpha1
+kind: InfrastructureProvider
+metadata:
+ name: azure
+ namespace: capz-system
+spec:
+ version: v0.4.9
+ secretName: azure-variables
+ fetchConfig:
+   selector:
+     matchLabels:
+       provider-components: azure
+```
+
+### Risks and Mitigation
+
+#### Error Handling & Logging
+
+Currently, clusterctl provides quick feedback regarding required variables
+etc. With the operator in place we’ll need to ensure that the error messages
+and logs are easily available to the user to verify progress.
+
+#### Extensibility Options
+
+Currently, clusterctl has a few extensibility options.  For example,
+clusterctl is built on-top of a library that can be leveraged to build other
+tools.
+
+It also exposes an interface for template processing if we choose to go a
+different route from `envsubst`. This may prove to be challenging in the
+context of the operator as this would mean a change to the operator
+binary/image. We could introduce a new behavior or communication protocol or
+hooks for the operator to interact with the custom template processor. This
+could be configured similarly to the fetch config, with multiple options built
+in.
+
+#### Upgrade from v1alpha3 management cluster to v1apha4/operator cluster
+
+Cluster API will provide instructions on how to upgrade from a v1alpha3
+management cluster, created by clusterctl to the new v1alpha4 management
+cluster. These operations could require manual actions.
+
+Some of the actions are described below:
+- Run webhooks as part of the main manager. See [issue
+  3822](https://github.com/kubernetes-sigs/cluster-api/issues/3822)
+
+## Additional Details
+
+### Test Plan
```

Comment:
II think we should explicitly talk about e2e tests.
- existing E2E test will verify the new version of init and upgrade as soon as ready
- a new E2E test is required to verify v1alpha3 --> v1alpha4 upgrades